pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -0,0 +1,389 @@
+	resyncPeriod time.Duration,
+	nodeName string,
+	podCache cnipodcache.CNIPodInfoStore,
+	cniServer *cniserver.CNIServer) *PodController {
+
+	pc := PodController{
+		kubeClient:         kubeClient,
+		netAttachDefClient: netAttachDefClient,
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "podcontroller"),
+		podInformer:        podInformer,
+		podLister:          corelisters.NewPodLister(podInformer.GetIndexer()),
+		nodeName:           nodeName,
+		podCache:           podCache,
+		cniServer:          cniServer,
+	}
+
+	podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    pc.enqueuePod,
+			DeleteFunc: pc.enqueuePod,
+			UpdateFunc: pc.enqueuePodUpdate,
+		},
+		resyncPeriod,
+	)
+	return &pc
+}
+
+func podKeyGet(pod *corev1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func (pc *PodController) enqueuePod(obj interface{}) {
+	var err error
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		podDeletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(err, "Unexpected object received:", obj)
+			return
+		}
+		pod, ok := podDeletedState.Obj.(*corev1.Pod)
+		if !ok {
+			klog.ErrorS(err, "DeletedFinalStateUnknown object is not of type Pod: ", podDeletedState.Obj, pod)
+			return
+		}
+	}
+	podKey := podKeyGet(pod)
+	pc.queue.Add(podKey)
+}
+
+func (pc *PodController) enqueuePodUpdate(prevObj, newObj interface{}) {
+	return
+}
+
+// handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
+func (pc *PodController) handleAddUpdatePod(obj interface{}) error {
+	var ret bool
+	pod := obj.(*corev1.Pod)
+
+	secondaryNetwork, ret := checkForPodSecondaryNetworkAttachement(pod)
+	if !ret {
+		klog.InfoS("Pod does not have a NetworkAttachmentDefinition", "Pod", klog.KObj(pod))
+		return nil
+	}
+	// get Pod/Container information (container id, namespace, etc.,) from CNI info cache (added by CNI Server CmdAdd upon
+	// successful Pod and primary network creation).
+	// If container information is not available yet in the CNICache, return error to requeue work
+	podCNIInfo := pc.podCache.GetCNIConfigInfoPerPod(pod.Name, pod.Namespace)
+	if len(podCNIInfo) == 0 {
+		return errors.New("CNI Cache Information not found")
+	} else {
+		networklist, err := parsePodSecondaryNetworkAnnotation(secondaryNetwork)
+		if err != nil {
+			return errors.New("secondary network annotation parsing failed")
+		}
+		pc.configureSecondaryNetwork(pod, networklist, podCNIInfo)
+	}
+	return nil
+}
+
+func (pc *PodController) handleRemovePod(key string) error {
+	var cmdArgs *invoke.Args
+
+	pod := strings.Split(key, "/")
+	// Read the CNI info (stored during Pod creation by cniserver) from cache.
+	// Delete CNI info shared in cache for a specific Pod which is getting removed/deleted.
+	podCNIInfo := pc.podCache.GetCNIConfigInfoPerPod(pod[1], pod[0])
```

Comment:
If CNI Del is handled before the Pod deletion event, how we keep the information in cache? Probably we should handle it at CNI Del too?

Reply:
From our offline discussion, I now get CNI Del will not delete the cached entry, but only podwatch will delete it when handling the Pod deletion event.
However, it can lead to another problem - if the Pod is deleted and recreated, in a race condition it is possible cniserver can create two entries for the old and new Pods, and GetCNIConfigInfoPerPod() can return both of them. @antoninbas : do you remember how we handle this inconsistence between CNI calls and Pod events?

Reply:
this is handled now as below:
1) at CNI server, before add CNI info, do a lookup on the cache for that specific Pod. if retrieved, delete the existing copy and add the new one. (no duplicates)
2) at podWatch, check for Status.PodIP and handle event only if this is set.